File: prototype/rails/src/app/actions/carts/current_cart.rb
module Carts
  class CurrentCart
    # A merge hands the verified customer whatever cart the anonymous visitor
    # was filling, so one customer can own two. The one holding the most items
    # is the one they were shopping with.
    def call(customer:)
      customer.carts.includes(:items).max_by { |cart| [cart.items.size, cart.id] } || customer.carts.create!
    end
  end
end

File: prototype/php/src/app/Actions/Cart/CurrentCart.php
<?php

declare(strict_types=1);

namespace App\Actions\Cart;

use App\Models\Cart;
use App\Models\Customer;

final class CurrentCart
{
    /**
     * A merge hands the verified customer whatever cart the anonymous visitor
     * was filling, so they can own two. The one holding items is the one the
     * visitor was shopping with.
     */
    public function __invoke(Customer $customer): Cart
    {
        return Cart::query()
            ->where('customer_id', $customer->id)
            ->withCount('items')
            ->orderByDesc('items_count')
            ->orderByDesc('id')
            ->first()
            ?? Cart::create(['customer_id' => $customer->id]);
    }
}
